Phi3: fix attn for sliding window

zucchini-nlp opened the pull request:
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/32945. The reason is that Phi3 prev prepared 4D attn with sliding window while the new `updte_causal_mask` didn't take that into account. This PR fixes it and adds a test

zucchini-nlp committed:
fix phi3 attn fir sliding window

zucchini-nlp commented on src/transformers/models/phi3/modeling_phi3.py:
```diff
@@ -88,25 +89,35 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
         causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
         if attention_mask is not None:
             causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+            if attention_mask.shape[-1] > target_length:
+                attention_mask = attention_mask[:, :target_length]
```
This was needed in case the input prompt is longer than sliding window, so we have to crop attention mask to the max possible length before masking out pad tokens

ArthurZucker replied:
Mmmm I don't think we can crop the attention mask for `eager`

zucchini-nlp replied:
For eager we never go beyond cache length because the `target_length` is `attention mask.shape[1]` whenever attention mask is present. So cropping shouldn't yield any difference

Same for static cache, where we expect cache shape to be always more than attention mask

ArthurZucker replied:
ah okay my bad then!

zucchini-nlp committed:
fix tests

gante commented on src/transformers/models/phi3/modeling_phi3.py:
```diff
@@ -1105,21 +1115,30 @@ def _update_causal_mask(
         min_dtype = torch.finfo(dtype).min
         sequence_length = input_tensor.shape[1]
-        if using_static_cache:
+        if using_sliding_window_cache:
```
we now have a significant code block to determine `target_length`, but `target_length` is not used directly in `_update_causal_mask`

suggestion to avoid this disconnection:
1. make `SlidingWindowCache` store the `config.sliding_window` it receives at init time
2. move the logic that computes `target_length` inside `_prepare_4d_causal_attention_mask_with_cache_position`
3. `_prepare_4d_causal_attention_mask_with_cache_position` no longer receives `target_length` nor `config`, as they can be retrieved from `past_key_values`

zucchini-nlp replied:
For the target length, maybe we merge https://github.com/huggingface/transformers/pull/32421/ first where the max length should be easily accessible through cache no matter cache type. We earlier discussed returning max cache shape, and sliding window has a max capacity for cache class

And then yeah, we can move that part to `_prepare_4d_causal_attention_mask_with_cache_position` but I think we'd better move it in all classes for general consistency

gante replied:
Agreed with the plan -- shall we leave a TODO for us in this PR, linking to your comment?

zucchini-nlp replied:
yes, I'll add a TODO for us so we can make the required change in all models at once, to not mix different updates in one PR :)

gante commented on src/transformers/models/phi3/modeling_phi3.py:
```diff
@@ -1364,15 +1385,25 @@ def prepare_inputs_for_generation(
             dtype = self.lm_head.weight.dtype
             min_dtype = torch.finfo(dtype).min
 
+            if isinstance(past_key_values, SlidingWindowCache):
```
(same comment as above here)

gante commented on src/transformers/models/phi3/modeling_phi3.py:
```diff
@@ -88,25 +89,35 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
-        if sequence_length != 1:
-            causal_mask = torch.triu(causal_mask, diagonal=1)
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+        exclude_mask = torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
```
A comment explaining what `exclude_mask` is doing would be nice to quickly understand this logic

I'm assuming it further zeroes out values in `causal_mask` that should be ignored because of the sliding window. But, at a first glance, I'm not understanding how (more descriptive variable names and storing intermediary tensors with good variable names might help with readbility)

gante commented on tests/models/phi3/test_modeling_phi3.py:
```diff
@@ -608,3 +608,89 @@ def test_phi3_mini_128k_instruct_with_static_cache(self):
 
         self.assertListEqual(output_text, EXPECTED_OUTPUT)
+
+    def test_phi3_mini_4k_sliding_window(self):
```
Tests 💛 

missing: a small comment at the top, explaining what this test should be testing (I'm assuming it tests the ability of the model to handle large sequence lengths with a sliding window, and that it is a regression test)

gante reviewed (commented):
Thank you for opening the PR with the fix 💪 A few questions and comments

zucchini-nlp commented on src/transformers/models/phi3/modeling_phi3.py:
```diff
@@ -88,25 +89,35 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
+        exclude_mask = torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+        if config.sliding_window is not None:
+            if not isinstance(past_key_values, SlidingWindowCache) or sequence_length > config.sliding_window:
+                exclude_mask.bitwise_or_(
+                    torch.arange(target_length, device=device)
+                    <= (cache_position.reshape(-1, 1) - config.sliding_window)
```
hehe, this was blindly copied from mistral. We can do it with "|" sign, but the current one is in-place so seems assignment is redundant a bit

gante replied:
I completely forgot that functions ending in `_` are the original operation but in-place!

(operations like `+=`, `*=`, ... are also in place and, in my opinion, more readable :p )

ArthurZucker reviewed (commented):
awesome test!

vasqu commented on tests/models/phi3/test_modeling_phi3.py:
```diff
@@ -608,3 +608,89 @@ def test_phi3_mini_128k_instruct_with_static_cache(self):
+        EXPECTED_OUTPUT = [
+            '1. Coq au Vin: Coq au Vin is a classic French dish that translates to "rooster in wine." The dish consists of chicken braised with wine, lardons, mushrooms, and garlic. It is a hearty and flavorful dish that is often served with potatoes or rice.\n\n            2. Boeuf Bourguignon: Boeuf Bourguignon is a traditional French beef stew that'
+        ]
+        print(output_text)
```
Nit: Ig this print is to be removed ;)

zucchini-nlp committed:
merge main

zucchini-nlp commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -1112,9 +1164,42 @@ def prepare_inputs_for_generation(
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and cache_position[0] == 0:
-            model_inputs = {"inputs_embeds": inputs_embeds}
+            model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
```
cant do that, phi3 has special treatment for long-short scales in RoPE, it was added recently and will be here after I rebase main

zucchini-nlp committed:
address most comment

zucchini-nlp committed:
style

zucchini-nlp commented:
Done, I think this PR is ready to be reviewed/merged.

Depending on which PR is merged first, this or the linked one, I will rebase and apply necessary changes. Then I'll add the TODO comment for moving `target_length` logic to `_prepare_attention_mask`

ArthurZucker commented on tests/models/phi3/test_modeling_phi3.py:
this LGTM! thanks for the thorough test

ArthurZucker commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -1118,9 +1170,42 @@ def prepare_inputs_for_generation(
+            dtype = self.lm_head.weight.dtype
+            min_dtype = torch.finfo(dtype).min
+
+            if isinstance(past_key_values, SlidingWindowCache):
+                sliding_window = (
+                    self.config.sliding_window if self.config.sliding_window is not None else sequence_length
+                )
+                target_length = max(sequence_length, sliding_window)
+            else:
+                target_length = past_key_values.get_max_length()
```
I don't understand why we need to do this: if there is a sliding WindowCache it was init from the `config` and thus has a correct `sliding_window`. 

Then, `SlidingWindowCache. get_max_length` should take `sequence_length` as input to return the max and avoid having these checks here WDYT?

zucchini-nlp replied:
Yes, it should but the PR for getting `max_length` on slidingWindow cache is on its way and not merged yet. 

So, to update you on our discussions with @gante which are currently in different PR comments: some cache classes now do not have a `max_length` (e.g. SlidingWindow). As commented in the code, sliding window technically has no max length and goes on a rolling basis. But in transformers what we want to check is the "maximum capacity of cache instance", independently of how cache handles new tokens going beyond that capacity.

So, in a different PR I changed naming to `get_max_cache_shape` which is more straightforward and added `get_max_cache_shape` for Sliding Window cache. We'll do a simple deprecation cycle, as we did for static cache's "max_batch_size". Until the linked PR is merged, I am copying this piece of code from `mistral` and using it in `phi3`. I have it noted and will handle it depending which one gets merged first :)

ArthurZucker replied:
alright sounds good. I just don't want us to add too much complexity to the code! 🤗

ArthurZucker reviewed (commented):
Thanks, let's try to abstract a tad bit, as we generally would avoid differentiating cache classes in the modeling!

gante commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -58,6 +58,72 @@
 _CONFIG_FOR_DOC = "MistralConfig"
 
 
+def _prepare_4d_causal_attention_mask_with_cache_position(
```
Because of #33677, this function is part of the model class -- I think you will have to move the diff there, otherwise tests may fail on `main`

(see the diff in that PR for `Llama`, it should be similar to the changes you need to do here)

gante reviewed (commented):
LGTM, but the changes will likely fail because of a recently merged PR (things need to be moved, see comment)

zucchini-nlp committed:
merge main

zucchini-nlp committed:
update after rebase

zucchini-nlp commented:
Rebased main and updated accordingly by moving `prepare_causal_mask` to XXXModel. Also, noticed Phi3Moe was added while PR was in progress and it is same as Phi3, so I propagated changes there too

Will be merging tomorrow if no comment remain :)

Cyrilvallez commented:
Hey @zucchini-nlp, while working on https://github.com/huggingface/transformers/pull/33619 I had issues with the 4d masks and just found this PR - however, it is not only an issue for Phi3: Mimi, Mixtral, PhiMoe, Qwen2, Qwen2Moe, Qwen2VL, Starcoder2 have the exact same issue (`AttentionMaskConverter._ignore_causal_mask_sdpa()` does not check for the `sliding_window`). 
Let me know if you can fix it or if you want me to jump on it.

zucchini-nlp commented:
@Cyrilvallez oh I see, didn't know we had more models that support sliding window. I can propagate changes to other models, sure :)

Cyrilvallez commented:
Yes! I think I listed them all but you can maybe double-check so that all of them get correctly fixed 🤗

ArthurZucker reviewed (commented):
ping me when merge this way I can put it in a patch!

zucchini-nlp committed:
add more models

zucchini-nlp committed:
merge main

zucchini-nlp committed:
fix tests

zucchini-nlp commented:
@ArthurZucker done! I had to change tests for Qwen2 models because otherwise we won't get same results for long padded input as for the base input. Applying sliding mask results in minor differences

ArthurZucker reviewed (approved):
Thanks, good for the test modifications, if run-slow were good let's go! 🔥
